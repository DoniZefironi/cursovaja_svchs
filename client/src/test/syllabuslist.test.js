import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SyllabusList from '../components/syllabuslist'; // Убедитесь, что путь правильный
import { FaFileDownload, FaEdit, FaTrash, FaPlus, FaMinus } from 'react-icons/fa';

const syllabuses = [
  { id: 1, name: 'Syllabus 1', date: '2023-01-15', syllfile: 'file1.pdf' },
  { id: 2, name: 'Syllabus 2', date: '2023-02-20', syllfile: 'file2.pdf' }
];

const currentUser = { user: { roles: ['ADMIN'] } };

const handleEdit = jest.fn();
const handleDelete = jest.fn();
const handleToggle = jest.fn();
const handleDownload = jest.fn();
const setCurrentSyllabus = jest.fn();
const setShowModal = jest.fn();
const setEditMode = jest.fn();

describe('SyllabusList Component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with correct syllabus information', () => {
    const { getByText, queryByText } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка наличия расписаний
    expect(getByText('Syllabus 1')).toBeInTheDocument();
    expect(getByText('15 Jan 2023')).toBeInTheDocument();
    expect(getByText('Syllabus 2')).toBeInTheDocument();
    expect(getByText('20 Feb 2023')).toBeInTheDocument();

    // Проверка наличия кнопок
    expect(queryByText('No syllabuses available')).not.toBeInTheDocument();
  });

  it('handles toggle correctly', () => {
    const { getAllByRole } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка вызова handleToggle
    const buttons = getAllByRole('button');
    fireEvent.click(buttons[buttons.length - 1]);
    expect(handleToggle).toHaveBeenCalledWith(1);
  });

  it('shows and hides additional information on toggle', () => {
    const { getByText, queryByText, rerender } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка начального состояния
    expect(queryByText('file1.pdf')).not.toBeInTheDocument();
    expect(queryByText('file2.pdf')).not.toBeInTheDocument();

    // Обновляем рендеринг компонента с openIndex = 0
    rerender(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={0}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка состояния после обновления
    expect(getByText('file1.pdf')).toBeInTheDocument();
    expect(queryByText('file2.pdf')).not.toBeInTheDocument();
  });

  it('handles download correctly', () => {
    const { getAllByRole } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка вызова handleDownload
    const downloadButton = getAllByRole('button', { name: /download/i })[0];
    fireEvent.click(downloadButton);
    expect(handleDownload).toHaveBeenCalledWith('file1.pdf');
  });

  it('handles edit and delete correctly', () => {
    const { getAllByRole } = render(
      <SyllabusList
        syllabuses={syllabuses}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentSyllabus={setCurrentSyllabus}
        setShowModal={setShowModal}
        setEditMode={setEditMode}
        handleDownload={handleDownload}
        handleDelete={handleDelete}
      />
    );

    // Проверка вызова handleEdit
    const editButton = getAllByRole('button', { name: /edit/i })[0];
    fireEvent.click(editButton);
    expect(setCurrentSyllabus).toHaveBeenCalledWith(syllabuses[0]);
    expect(setShowModal).toHaveBeenCalledWith(true);
    expect(setEditMode).toHaveBeenCalledWith(true);

    // Проверка вызова handleDelete
    const deleteButton = getAllByRole('button', { name: /delete/i })[0];
    fireEvent.click(deleteButton);
    expect(handleDelete).toHaveBeenCalledWith(1);
  });
});
