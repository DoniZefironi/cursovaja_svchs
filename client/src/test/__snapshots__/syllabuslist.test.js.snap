// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyllabusList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content-box list-box width100"
  >
    <div
      class="contentiks"
    >
      <div
        class="list-item"
      >
        <div>
          <h6>
            Syllabus 1
          </h6>
          <p>
            15 Jan 2023
          </p>
        </div>
        <div
          class="iconochke"
        >
          <div>
            <button
              class="btn btn-link"
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 384 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm76.45 211.36l-96.42 95.7c-6.65 6.61-17.39 6.61-24.04 0l-96.42-95.7C73.42 337.29 80.54 320 94.82 320H160v-80c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v80h65.18c14.28 0 21.4 17.29 11.27 27.36zM377 105L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128v-6.1c0-6.3-2.5-12.4-7-16.9z"
                />
              </svg>
            </button>
          </div>
          <button
            class="mx-2 btn btn-warning"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              />
            </svg>
          </button>
          <button
            class="btn btn-danger"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
          <button
            class="plus-button btn btn-light"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="additional-info collapse"
      >
        <p>
          file1.pdf
        </p>
      </div>
    </div>
    <div
      class="contentiks"
    >
      <div
        class="list-item"
      >
        <div>
          <h6>
            Syllabus 2
          </h6>
          <p>
            20 Feb 2023
          </p>
        </div>
        <div
          class="iconochke"
        >
          <div>
            <button
              class="btn btn-link"
              type="button"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 384 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm76.45 211.36l-96.42 95.7c-6.65 6.61-17.39 6.61-24.04 0l-96.42-95.7C73.42 337.29 80.54 320 94.82 320H160v-80c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v80h65.18c14.28 0 21.4 17.29 11.27 27.36zM377 105L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128v-6.1c0-6.3-2.5-12.4-7-16.9z"
                />
              </svg>
            </button>
          </div>
          <button
            class="mx-2 btn btn-warning"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              />
            </svg>
          </button>
          <button
            class="btn btn-danger"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              />
            </svg>
          </button>
          <button
            class="plus-button btn btn-light"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="additional-info collapse"
      >
        <p>
          file2.pdf
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
